runtime/src/crowdfund.rs

```diff
@@ -0,0 +1,539 @@
+			let who = ensure_signed(origin)?;
+
+			ensure!(value >= T::MinContribution::get(), "contribution too small");
+
+			let mut fund = Self::funds(index).ok_or("invalid fund index")?;
+			fund.raised += value;
+			ensure!(fund.raised <= fund.cap, "contributions exceed cap");
+			let now = <system::Module<T>>::block_number();
+			ensure!(fund.end > now, "contribution period ended");
+
+			T::Currency::transfer(&who, &Self::fund_account_id(index), value)?;
+
+			let id = Self::id_from_index(index);
+			let balance = who.using_encoded(|b| child::get_or_default::<BalanceOf<T>>(id.as_ref(), b));
+			let balance = balance.saturating_add(value);
+			who.using_encoded(|b| child::put(id.as_ref(), b, &balance));
+
+			if <slots::Module<T>>::is_ending(now).is_some() {
+				// Now in end period; record it
+				fund.last_contribution = Some(now);
+				if let Some(c) = fund.last_contribution {
+					if c != now {
+						// last contribution was at earlier time; re-insert into `NewRaise`
+						NewRaise::mutate(|v| v.push(index));
+					}
+				}
+			} else {
+				if fund.last_contribution.is_some() {
+					// Now outside end period and last was also inside end period: reset last to
+					// None and insert (because it will have been removed at end of last block).
+					fund.last_contribution = None;
+					NewRaise::mutate(|v| v.push(index));
+				}
+			}
+
+			<Funds<T>>::insert(index, &fund);
+
+			Self::deposit_event(RawEvent::Contributed(who, index, value));
+		}
+
+		
+		/// Withdraw full balance of a contributor to an unsuccessful fund.
+		fn withdraw(origin, #[compact] index: FundIndex) {
+			let who = ensure_signed(origin)?;
+
+			let mut fund = Self::funds(index).ok_or("invalid fund index")?;
+			let now = <system::Module<T>>::block_number();
+			ensure!(now >= fund.end, "contribution period not over");
+
+			let id = Self::id_from_index(index);
+			let balance = who.using_encoded(|b| child::get::<BalanceOf<T>>(id.as_ref(), b))
+				.ok_or("not a contributor")?;
+
+			// Avoid using transfer to ensure we don't pay any fees.
+			let _ = T::Currency::resolve_into_existing(&who, T::Currency::withdraw(
+				&Self::fund_account_id(index),
+				balance,
+				WithdrawReason::Transfer,
+				ExistenceRequirement::AllowDeath
+			)?);
+
+			let id = Self::id_from_index(index);
+			who.using_encoded(|b| child::kill(id.as_ref(), b));
+			fund.raised = fund.raised.saturating_sub(balance);
+
+			<Funds<T>>::insert(index, &fund);
+
+			Self::deposit_event(RawEvent::Withdrew(who, index, balance));
+		}
+		
+		/// Note that a successful fund has lost its parachain slot, and place it into retirement.
+		fn begin_retirement(_origin, #[compact] index: FundIndex) {
+			// origin unimportant.
+			let mut fund = Self::funds(index).ok_or("invalid fund index")?;
+			let _parachain_id = fund.parachain.take().ok_or("fund has no parachain")?;
+			let account = Self::fund_account_id(index);
+			ensure!(T::Currency::free_balance(&account) >= fund.raised, "funds not yet returned");
```

Comment:
This seems to be the only "check" that this function can be called to end a crowdfund.
However, it is not clear yet that someone will not be able to call this:
* When the crowdfund has collected all the funds
* Before `handle_bid` is called on the fund, reserving them

It seems we imply that a fund having a `para_id` means that it has completed onboarding, but that does not seem to be the full story...

Reply:
Having finished reading the full code and understanding to the best of my ability, I think this comment still stands.

Reply:
> It seems we imply that a fund having a para_id means that it has completed onboarding

correct. what's the rest of the story?

Reply:
I think I misunderstood, and it does seem right. The story:

* Parachain opens crowdfund, collects as much funds as possible up to cap. No parachain id assigned to it.
* When auction `is_ending` a bid is sent from the crowdfund with the total amount raised so far, reducing its free balance.
* A crowdfund can finally win the bid, and someone will call `onboard` which will ultimately assign it a para_id, but only if the onboarding process succeeds in `slots` too.

At no point will a fund have both its total funds raised and a para_id until after it has lost its slot.

Thanks!